### ISSUE TITLE
Add noEmptyLinesBeforeNonFirstDecls option to declaration-block-properties-order

### ISSUE THREAD

**Issue by dan-gamble** [stylelint/stylelint#752]
Replicated at: https://github.com/dan-gamble/styletest/tree/master

Give then `rule-properties-order` of:

```
'rule-properties-order': [
      {
        emptyLineBefore: true,
        properties: [
          'content'
        ]
      },
      {
        emptyLineBefore: true,
        properties: [
          'position',
          'top',
          'right',
          'bottom',
          'left',
          'z-index'
        ]
      },
      {
        emptyLineBefore: true,
        properties: [
          'display',
          'float',
          'height',
          'margin',
          'margin-top',
          'margin-right',
          'margin-bottom',
          'margin-left',
          'max-height',
          'max-width',
          'min-height',
          'min-width',
          'padding',
          'padding-top',
          'padding-right',
          'padding-bottom',
          'padding-left',
          'table-layout',
          'width'
        ]
      }
    ]
```

This CSS passes:

```
.sec {
  content: 'tree';
  
  position: absolute;
  
  top: 0;
  right: 0;
  
  bottom: 0;
  
  left: 0;
}
```

There shouldn't be any empty lines between properties in the same group.

**Comment by davidtheclark** [comment]
I'm not sure whether to consider this a bug or an enhancement. @jeddy3 ?

**Comment by dan-gamble** [comment]
I'd personally err on the side of bug because having the rule `emptyLineBefore` implies to me that's the only place an empty line should be and in the docs the only time an empty line before appears is when that option is used. Just my 2c though

**Comment by jeddy3** [comment]
I think it should be an enhancement and enabled by an explicit option (e.g. `noEmptyLines: true`), as:

1. We don't restrict empty lines above declarations else where.
2. It's going to conflict with [`comment-empty-line-before`](http://stylelint.io/user-guide/rules/comment-empty-line-before/). For example:

```css
a {
  content: "foo";

  position: absolute;
  z-index: 2;
  
  display: block;

  /* some comment about the next two declarations .e.g. "Match size to image" */
  height: 100px;
  width: 100px;
}

```

I think users need to opt-in into that kind of behaviour.

**Comment by dan-gamble** [comment]
You guys know better than me, i'm game for that :+1: 

Just realised it allows me to do the comment thing in it's current state so a switch does sound like the better way to go.

How would comments be handled if this was switched on?

**Comment by davidtheclark** [comment]
> How would comments be handled if this was switched on?

That would be a problem. You might end up with rules contradicting each other. @jeddy3, have we managed to completely avoid that so far, or do we have precedent for dealing with hit?

**Comment by jeddy3** [comment]
> have we managed to completely avoid that so far

Yep, we've always kept rules standalone and maintained that resolving conflicts is the responsibility of the user.

> How would comments be handled if this was switched on?

I think we can nip this one in the bud (and avoid the potential for conflicts) by renaming the option to `noEmptyLinesBeforeNonFirstDecls: true`. ("NonFirst" is weird though, any better suggestions?). We can then reuse the [empty line logic](https://github.com/stylelint/stylelint/blob/master/src/rules/rule-properties-order/index.js#L215) already within the rule. It also means that there will be *no conflict with* `comment-empty-line-before` as both these...:

`comment-empty-line-before: "never"`:

```css
a {
  content: "foo";

  position: absolute;
  z-index: 2;

  display: block;
  /* some comment about the next two declarations .e.g. "Match size to image" */
  height: 100px;
  width: 100px;
}
```

`comment-empty-line-before: "always"`:

```css
a {
  content: "foo";

  position: absolute;
  z-index: 2;

  display: block;

  /* some comment about the next two declarations .e.g. "Match size to image" */
  height: 100px;
  width: 100px;
}
```

... are valid as far as `noEmptyLinesBeforeNonFirstDecls: true` is concerned as each "non first" declaration is just checking if it's preceded by an empty time, which is never is.

**Comment by davidtheclark** [comment]
Sounds good. I'm open to a PR adding the `noEmptyLinesBeforeNonFirstDecls` option.

**Comment by dan-gamble** [comment]
I'll take my 3rd stab at this, hopefully it doesn't leave me crying like my last 2 attempts. This is my rule that i asked to be added so it's only fair the stress that comes with it be on my head :D

**Comment by davidtheclark** [comment]
@dan-gamble, sorry your attempts are on this, one of the most complex and frustrating rules. Maybe you should take a stab at something really easy to ease your pain and restore your ego?

**Comment by dan-gamble** [comment]
Haha it's not a bad shout, at the very least i'll look at the rule and see if i can do anything, will report back tonight hopefully :+1:

**Comment by dan-gamble** [comment]
Started a really untidy base to just try and see if i'm on the right lines https://github.com/dan-gamble/stylelint/commit/197ed2d0728bd100d74ad47c62195da05b68b957

**Comment by davidtheclark** [comment]
Sure seems like the right track. The most important part is just to get the right tests working without breaking the other tests, and that's what you're shooting for.

**Comment by davidtheclark** [comment]
@dan-gamble Is your work here also going to take care of https://github.com/stylelint/stylelint/issues/578?

**Comment by dan-gamble** [comment]
I could add another option that fixes that too. I imagine once i've cracked this, that would be fairly trivial. I imagine that would be fairly trivial in itself, of just reversing the function that checks for a new line anyway?

Work on this has been slow from my end recently as i bought a new car Sunday and have had countless problems with it so been trying to sort that.. Hopefully i can weigh in with some effort on it tonight

**Comment by davidtheclark** [comment]
Sorry to hear about the car. I looked back at that issue again and see that I misread something, and these are more separate concerns than I thought for some reason. Just do whatever you can. :)

**Comment by dan-gamble** [comment]
Yea, i could potentially look into #578 as i image it's a lot easier than this one. Will look into it tonight, if it's easier to get my head around i'll push that one out first :+1:

**Comment by dan-gamble** [comment]
I'm thinking after the PR for #578 i should be able to wrap this one up tomorrow now i have a better understand of how this rule is working :+1:

**Comment by evilebottnawi** [comment]
@jeddy3 i think we should close this issue with remark about deprecated objects groups and plugins.

**Comment by jeddy3** [comment]
I was going to wait until we're all in agreement in https://github.com/stylelint/stylelint/issues/1308 and that issue is marked as "enhancement".

**Comment by jeddy3** [comment]
See: https://github.com/stylelint/stylelint/issues/1308
